### ISSUE TITLE
Lint for unused arguments starting with a _

### ISSUE THREAD

**Issue by yaahallo** [rust-lang-nursery/rust-clippy#3219]
In response to [this comment](https://github.com/rust-lang-nursery/rust-clippy/pull/3203#discussion_r219992918)

I think the idea here would be to recommend removal of arguments that are ignored and never referenced.

On the flip side, I'm wondering if it would make sense to also lint for used arguments starting `_` which recommends removing the `_`, because that came up in the same review.

**Comment by yaahallo** [comment]
@flip1995

**Comment by flip1995** [comment]
Yes I would suggest to make the unused `_arg` lint a pedantic lint (it can actually make sense to have such an unused arg: [StackOverflow](https://stackoverflow.com/a/48370313)) and the used `_arg` lint a style lint.

**Comment by Zhonghe-zhao** [comment]
I'd like to work on this issue. Could I claim it?

**Comment by samueltardieu** [comment]
> I'd like to work on this issue. Could I claim it?

Sure!

**Comment by samueltardieu** [comment]
The scope is for *unused* parameters starting with `_`, not *used* parameters. Such as in

```rust
fn foo(a: u32, _b: _u32) {
    println!("a = {a}");
}
```

It should suggest removing `_b`, but it has to check that:
- `foo` is not conditionally defined through a `#[cfg]` attribute
- `_b` is really unused throughout the function, including in code conditionally compiled
- `foo` is not being exported, at least locally, through a function ref, as it might require keeping this extra parameter

and probably other use cases.

By the way, Clippy will enter a feature freeze mode in a few days, and new lints won't be accepted for about 3 months. If you start a new lint, and it is not ready to be merged before the middle of next week, then it might be blocked for quite a long time. If you change your mind, don't hesitate to release the assignment, and maybe work on fixing an existing issue (those won't be frozen, only new features will be blocked).

**Comment by Zhonghe-zhao** [comment]
> The scope is for _unused_ parameters starting with `_`, not _used_ parameters. Such as in
> 
> fn foo(a: u32, _b: _u32) {
>     println!("a = {a}");
> }
> It should suggest removing `_b`, but it has to check that:
> 
> * `foo` is not conditionally defined through a `#[cfg]` attribute
> * `_b` is really unused throughout the function, including in code conditionally compiled
> * `foo` is not being exported, at least locally, through a function ref, as it might require keeping this extra parameter
> 
> and probably other use cases.
> 
> By the way, Clippy will enter a feature freeze mode in a few days, and new lints won't be accepted for about 3 months. If you start a new lint, and it is not ready to be merged before the middle of next week, then it might be blocked for quite a long time. If you change your mind, don't hesitate to release the assignment, and maybe work on fixing an existing issue (those won't be frozen, only new features will be blocked).

Hi, thank you very much for taking the time to explain all the details and potential edge cases. I really appreciate the thoughtful guidance from all the Clippy contributors.

After reading your explanation, I realize that this lint is more complex than I initially thought, and I might not be able to complete it before the feature freeze. So I’ll unassign myself from this issue for now.

Instead, I’ll try to contribute by starting with smaller or existing issues, and gradually build up my understanding of Clippy and the compiler internals. I hope to come back and work on this lint when I’m better prepared and the timing is right.

Thanks again for your support!

**Comment by clueless-skywatcher** [comment]
Hi I would like to take this up
@rustbot claim
